libraries/datasource_okhttp/src/main/java/androidx/media3/datasource/okhttp/OkHttpDataSource.java

```diff
@@ -357,6 +361,10 @@ public int read(byte[] buffer, int offset, int length) throws HttpDataSourceExce
     try {
       return readInternal(buffer, offset, length);
     } catch (IOException e) {
+      if (e instanceof InterruptedIOException) {
+        closeConnectionQuietly();
```

Comment:
Thanks.  In that case I suspect there is another bug or unexpected codepath that is making that not reliably happen. This gives me something useful to test, and confirm.  If I can confirm that that close isn't happening reliably, that might be the better place to fix this.

In my manual testing this fix does appear to fix the SocketTimeoutException reports.  Interrupt logic in OkHttp is not widely used, so it definitely requires careful cleanup here. Based on my observation of the Media3 behaviour I was able to make a reproduction in the OkHttp project, and closing the Response after an InterruptedIOException is thrown, resolves that, hence bringing that fix back into this project, and likewise it appears to fix it.

Reply:
If you can reproduce a case where `open` is called on a `DataSource` and a corresponding `close` call doesn't happen after a read failure, then yes we should fix that instead.

Reply:
I tested briefly with an AtomicInteger incremented after call.execute, and decremented on response.close, but it confirmed your assertion that it's closed reliably.  

When it fails, it gets wedged with the server stopping sending additional data. OkHttp Frame logging shows data frame cease, and new requests don't make any progress, either because the stream/connection flow control window is depleted, or because the interrupted responses aren't closed.

It's difficult to replicate this cleanly manually in the sample app.  I might end up making a test case using the MediaBrowser to control predictably in a test, and use that to get it failing.